src/qibo/quantum_info/quantum_networks.py

```diff
@@ -629,53 +565,497 @@ def _run_checks(self, partition, system_output, pure):
+            the Einstein summation convention. For more details, please refer to
+            `numpy.einsum <https://numpy.org/doc/stable/reference/generated/numpy.einsum.html>`_.
+        operands (:class:`qibo.quantum_info.quantum_networks.QuantumNetwork`): Quantum
+            networks to be contracted.
+
+    Returns:
+        :class:`qibo.quantum_info.quantum_networks.QuantumNetwork`: Quantum network resulting
+            from the link product between two quantum networks.
+    """
+
+    if not isinstance(subscripts, str):
+        raise_error(
+            TypeError,
+            f"subscripts must be type str, but it is type {type(subscripts)}.",
         )
+
+    for i, operand in enumerate(operands):
+        if not isinstance(operand, QuantumNetwork):
+            raise_error(TypeError, f"The {i}th operator is not a ``QuantumNetwork``.")
+
+    if backend is None:  # pragma: no cover
+        backend = operands[0]._backend
+
+    tensors = [
+        (
+            backend.to_numpy(operand.full())
+            if operand.is_pure()
+            else backend.to_numpy(operand._tensor)
+        )
+        for operand in operands
+    ]
+
+    # keep track of the `partition` and `system_input` of the network
+    _, contracrtion_list = np.einsum_path(
+        subscripts, *tensors, optimize=False, einsum_call=True
+    )
+
+    inds, idx_rm, einsum_str, remaining, blas = contracrtion_list[0]
+    input_str, results_index = einsum_str.split("->")
+    inputs = input_str.split(",")
+
+    # Warning if the same index connects two input or two output systems
+    if not surpress_warning:
+        for ind in idx_rm:
+            found = 0
+            for i, script in enumerate(inputs):
+                try:
+                    index = script.index(ind)
+                    found += 1
+                    if found > 1:
+                        if is_input != operands[inds[i]].system_input[index]:
+                            pass
+                        else:
+                            warning(
+                                f"Index {ind} connects two {'input' if is_input else 'output'} systems."
+                            )
+                    is_input = operands[inds[i]].system_input[index]
+                    if found > 2:
+                        warning(
+                            f"Index {ind} appears multiple times in the input subscripts {input_str}."
+                        )
+                except:
+                    continue
+
+    # set correct order of the `partition` and `system_input`
+    partition = []
+    system_input = []
+    for ind in results_index:
+        for i, script in enumerate(inputs):
+            try:
+                index = script.index(ind)
+
+                partition.append(operands[inds[i]].partition[index])
+                system_input.append(operands[inds[i]].system_input[index])
+
+            except:
+                continue
```

Comment:
Same here

Reply:
Ah, this is for checking the index of each letter. As the letter may not be in the script, in which case nothing needs to be done, I simply suppress the error for `script.index(ind)`